ssl/quic/quic_impl.c

```diff
@@ -4623,6 +4632,261 @@ static QUIC_CONNECTION *create_qc_from_incoming_conn(QUIC_LISTENER *ql, QUIC_CHA
+ * tokencache
+ *
+ * In an effort to make allocation and freeing of these tokens a bit faster
+ * We do them in a single allocation in this format
+ * +---------------+        --\
+ * |   hashkey *   |---|      |
+ * |   hashkey_len |   |      | QUIC_TOKEN
+ * |   token *     |---|--|   |
+ * |   token_len   |   |  |   |
+ * +---------------+<--|  | --/
+ * |  hashkey buf  |      |
+ * |               |      |
+ * |---------------|<-----|
+ * |  token buf    |
+ * |               |
+ * +---------------+
+ *
+ * @param peer - the peer address that sent the token
+ * @param token - the buffer holding the token
+ * @param token_len - the size of token
+ *
+ * @returns a QUIC_TOKEN pointer or NULL on error
+ */
+static QUIC_TOKEN *ossl_quic_build_new_token(BIO_ADDR *peer, uint8_t *token,
+                                             size_t token_len)
+{
+    QUIC_TOKEN *new_token;
+    size_t hashkey_len = 0;
+    size_t addr_len = 0;
+    int family;
+    unsigned short port;
+    int *famptr;
+    unsigned short *portptr;
+    uint8_t *addrptr;
+
+    if ((token != NULL && token_len == 0) || (token == NULL && token_len != 0))
+        return NULL;
+
+    if (!BIO_ADDR_rawaddress(peer, NULL, &addr_len))
+        return NULL;
+    family = BIO_ADDR_family(peer);
+    port = BIO_ADDR_rawport(peer);
+
+    hashkey_len += sizeof(int); /* hashkey(family) */
+    hashkey_len += sizeof(unsigned short); /* hashkey(port) */
+    hashkey_len += addr_len; /* hashkey(address) */
+
+    new_token = OPENSSL_zalloc(sizeof(QUIC_TOKEN) + hashkey_len + token_len);
+    if (new_token == NULL)
+        return NULL;
+
+    if (!CRYPTO_NEW_REF(&new_token->references, 1)) {
+        OPENSSL_free(new_token);
+        return NULL;
+    }
+
+    new_token->hashkey_len = hashkey_len;
+    /* hashkey is allocated inline, immediately after the QUIC_TOKEN struct */
+    new_token->hashkey = (uint8_t *)(new_token + 1);
+    /* token buffer follows the hashkey in the inline allocation */
+    new_token->token = new_token->hashkey + hashkey_len;
+    new_token->token_len = token_len;
+    famptr = (int *)new_token->hashkey;
+    portptr = (unsigned short *)(famptr + 1);
+    addrptr = (uint8_t *)(portptr + 1);
+    *famptr = family;
+    *portptr = port;
+    if (!BIO_ADDR_rawaddress(peer, addrptr, NULL)) {
+        ossl_quic_free_peer_token(new_token);
+        return NULL;
+    }
+    if (token != NULL)
+        memcpy(new_token->token, token, token_len);
+    return new_token;
+}
+
+int ossl_quic_set_peer_token(SSL_CTX *ctx, BIO_ADDR *peer,
+                             const uint8_t *token, size_t token_len)
+{
+    SSL_TOKEN_STORE *c = ctx->tokencache;
+    QUIC_TOKEN *tok, *old = NULL;
+
+    if (ctx->tokencache == NULL)
+        return 0;
+
+    tok = ossl_quic_build_new_token(peer, (uint8_t *)token, token_len);
+    if (tok == NULL)
+        return 0;
+
+    /* we might be sharing this cache, lock it */
+    ossl_crypto_mutex_lock(c->mutex);
```

Comment:
Do the lock and unlock have to be guarded by `# if defined(OPENSSL_THREADS)`?

Reply:
no, ossl_crypto_mutex_[lock|unlock] are defined for the no-threads case, and return void, so their automatically noops in that event